### PR TITLE
added softStop method which waits until all batches uploaded then stop;

### PR DESCRIPTION
**Pull request by Megaprog**
renamed forceStop to hardStop;
AppMetrAsync uses softStop in stop() method and returns CF

### CODE REVIEW

File: src/main/java/com/appmetr/s2s/AppMetrAsync.java [hunk 1]
```diff
@@ -26,7 +26,10 @@ public AppMetrAsync(AppMetr appMetr, ScheduledExecutorService executorService) {
         this.appMetr = appMetr;
         this.executorService = executorService;
 
-        appMetr.start();
+        if (appMetr.isStopped()) {
```
Comment (AntonGabov) [inline src/main/java/com/appmetr/s2s/AppMetrAsync.java]: Is it correct logic? We should start `appMetr` when it's stopped

File: src/main/java/com/appmetr/s2s/AppMetrAsync.java [hunk 2]
```diff
@@ -47,12 +50,19 @@ public AppMetrAsync(AppMetr appMetr, ScheduledExecutorService executorService) {
...
-        appMetr.stop();
+    public CompletableFuture<Void> stop() {
+        return CompletableFuture.runAsync(() -> {
+            periodicFlushingFuture.cancel(false);
+            executorService.shutdown();
+        }, executorService).thenCompose(aVoid -> CompletableFuture.runAsync(() -> {
+            try {
+                awaitTermination();
+                appMetr.flush();
+                appMetr.softStop();
```
Comment (AntonGabov) [inline src/main/java/com/appmetr/s2s/AppMetrAsync.java]: in `appMetr.softStop()` you already do` flush()`.
Do we really need to call `flush()` before `stop()`?

File: src/main/java/com/appmetr/s2s/AppMetrAsync.java (hunk 1 above)
Comment (Megaprog) [reply src/main/java/com/appmetr/s2s/AppMetrAsync.java]: only 2 states are available stopped and not stopped. if it is not stopped then it it running and no need to start

File: src/main/java/com/appmetr/s2s/AppMetrAsync.java (hunk 2 above)
Comment (Megaprog) [reply src/main/java/com/appmetr/s2s/AppMetrAsync.java]: yes, we don't

File: src/main/java/com/appmetr/s2s/AppMetrAsync.java (hunk 1 above)
Comment (AntonGabov) [reply src/main/java/com/appmetr/s2s/AppMetrAsync.java]: But in your case: AppMetr is not stopped and not running, but you start it.

Comment (Megaprog) [reply src/main/java/com/appmetr/s2s/AppMetrAsync.java]: new Appmetr() is stopped until start() method invoked

**Pull request merged**